File: cses.fi/TreeDistances1.java
/*
 * https://cses.fi/problemset/task/1132
 */
import java.io.*;
import java.lang.*;
import java.util.*;


/*
 * Key Concept here -> when we do BFS from a particular node, then the last node obtained is 1 of the 
 * endpoints of the diameter of the tree.
 */
public class TreeDistances1 {
	
	private static int n;
	private static Map<Integer, List<Integer> > graph = new HashMap<>();
	
	private static int bfs1(int node) {
		int farthestNode = 0;
		Queue<Integer> q = new LinkedList<>();
		Set<Integer> vis = new HashSet<>();
		q.offer(node);
		while (!q.isEmpty()) {
			node = q.poll();
			if (vis.contains(node))
				continue;
			vis.add(node);
			farthestNode = node;
			for (int neighbor : graph.getOrDefault(node, Collections.emptyList())) {
				if (!vis.contains(neighbor))
					q.offer(neighbor);
			}
		}
		return farthestNode;
	}
	
	private static int[] bfs2(int node) {
		int[] dist = new int[n + 1];
		Queue<int[]> q = new LinkedList<>();
		q.offer(new int[] {node, 0});
		Set<Integer> vis = new HashSet<>();
		while (!q.isEmpty()) {
			int[] temp = q.poll();
			node = temp[0];
			int d = temp[1];
			if (vis.contains(node))
				continue;
			vis.add(node);
			dist[node] = d;
			for (int neighbor : graph.getOrDefault(node, Collections.emptyList())) {
				if (!vis.contains(neighbor))
					q.offer(new int[] {neighbor, d + 1});
			}
		}
		return dist;
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(br.readLine().trim());
		String[] temp;
		for (int i = 1; i < n; i++) {
			temp = br.readLine().trim().split(" ");
			int u = Integer.parseInt(temp[0]), v = Integer.parseInt(temp[1]);
			graph.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
			graph.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
		}
		int dep1 = bfs1(1), dep2 = bfs1(dep1);
		int[] dist1 = bfs2(dep1), dist2 = bfs2(dep2);
		for (int i = 1; i <= n; i++)
			System.out.print(Math.max(dist1[i], dist2[i]) + " ");
	}
}

File: cses.fi/TreeDistances1.cpp
/*
 * https://cses.fi/problemset/task/1132
 */
#include <bits/stdc++.h>
using namespace std;

int n;
vector<vector<int> > graph;

int bfs1(int node) {
	queue<int> q;
	int ans = 0;
	q.push(node);
	unordered_set<int> ust;
	while (!q.empty()) {
		node = q.front();
		q.pop();
		if (ust.find(node) != ust.end())
			continue;
		ust.insert(node);
		ans = node;
		for (int neighbor : graph[node]) {
			if (ust.find(neighbor) == ust.end())
				q.push(neighbor);
		}
	}
	return ans;
}

vector<int> bfs2(int node) {
	vector<int> v(n + 1, 0);
	queue<pair<int, int> > q;
	unordered_set<int> ust;
	q.push({node, 0});
	while (!q.empty()) {
		pair<int, int> p = q.front();
		q.pop();
		node = p.first;
		if (ust.find(node) != ust.end())
			continue;
		ust.insert(node);
		int d = p.second;
		v[node] = d;
		for (int neighbor : graph[node]) {
			if (ust.find(neighbor) == ust.end())
				q.push({neighbor, d + 1});
		}
	}
	return v;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	int u, v;
	cin >> n;
	graph.resize(n + 1);
	for (int i = 1; i < n; i++) {
		cin >> u >> v;
		graph[u].push_back(v);
		graph[v].push_back(u);
	}
	int dep1 = bfs1(1), dep2 = bfs1(dep1);
	vector<int> v1 = bfs2(dep1), v2 = bfs2(dep2);
	for (int i = 1; i <= n; i++)
		cout << max(v1[i], v2[i]) << ' ';
}
